File: 142. Linked_List_Cycle_II/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        
        map<ListNode*,int> history;
        
        ListNode* it = head;
        int idx_count = 0;
        while(it != nullptr){
            if(history.find(it) != history.end())
                return it;
            else{
                history[it] = idx_count;
                idx_count++;
            }
                
            it = it->next;
        }
        return nullptr;
        
    }
};

File: 142. Linked_List_Cycle_II/solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        
        ListNode slow = head;
        ListNode fast = head;
        ListNode enter = head;
        
        while(!Objects.isNull(fast) && !Objects.isNull(fast.next)){
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast){
                while(slow != enter){
                    enter = enter.next;
                    slow = slow.next;
                }
                return enter;
            }
        }
        
        return null;
        
    }
}